pyquil/quilatom.py

```diff
@@ -537,6 +537,51 @@ def __neg__(self) -> "Mul":
     def _substitute(self, d: Any) -> ExpressionDesignator:
         return self
 
+    def _evaluate(self) -> np.complex128:
+        """
+        Attempts to evaluate the expression to by simplifying it to a complex number.
+
+        Expression simplification can be slow, especially for large recursive expressions.
+        This method will raise a ValueError if the expression cannot be simplified to a complex
+        number.
+        """
+        expr = quil_rs_expr.Expression.parse(str(self))
+        expr.simplify()  # type: ignore
+        if not expr.is_number():
+            raise ValueError(f"Cannot evaluate expression {self} to a number. Got {expr}.")
+        return np.complex128(expr.to_number())
+
+    def __float__(self) -> float:
+        """
+        Returns a copy of the expression as a float by attempting to simplify the expression.
+
+        Expression simplification can be slow, especially for large recursive expressions.
+        This cast will raise a ValueError if simplification doesn't result in a real number.
+        """
+        value = self._evaluate()
+        if value.imag != 0:
+            raise ValueError(f"Cannot convert complex value with non-zero imaginary value to float: {value}")
+        return float(value.real)
+
+    def __array__(self, dtype: Optional[np.dtype] = None) -> np.ndarray:
+        """
+        Implements the numpy array protocol for this expression.
+
+        If the dtype is not object, then there will be an attempt to simplify the expression to a
+        complex number. If the expression cannot be simplified to one, then fallback to the
+        object representation of the expression.
+
+        Note that expression simplification can be slow for large recursive expressions.
+        """
+        try:
+            if dtype != object:
+                return np.asarray(self._evaluate(), dtype=dtype)
+            raise ValueError
+        except ValueError:
+            # Note: The `None` here is a placeholder for the expression in the numpy array.
+            # The expression instance will still be accessible in the array.
+            return np.array(None, dtype=object)
```

Comment:
@bramathon @steve-jeffrey @mhodson-rigetti 

Any thoughts on this proposed behavior? Do you see this causing more problems than it would solve? Is the potential performance cause of expression simplification going to be an issue?

Reply:
It doesn't seem ideal. Why isn't the `pyquil.quilatom.Expression` bound to the underlying Rust expression type so that it is built up in concert with the Python version, and can call directly on Rust to evaluate the existing expression?

Given this has only come up in the context of DEFGATE, there may not be a practical performance implication.

Reply:
Due to limitations with pyo3 and how the quil-py bindings are constructed, inheritance of its `Expression` class isn't possible. Since the pyQuil `Expression` leans heavily on inheritance for sharing functionality, parsing a string at the boundary is a tradeoff we made to avoid having to pursue a number of more hacky things to map them to quil-rs expressions. I agree that is less than ideal, and I'm hoping to reconstruct the compatibility layer by making more of pyQuil Rust native and depend directly on quil-rs, rather than try to bridge the gap with quil-py (#1755). Among other things, that approach won't have the limitation around subclassing, and we should be able to build the expressions up in concert.

All of that said, the behavior of how the array protocol works is unlikely to change. It won't require parsing in the future, but fundamentally it will attempt to simplify and evaluate the expression as described here. If that bit is okay, then this is resolves a pain point today, and adds tests to make sure it's carried forward as we continue to make things better.

Reply:
Understood. At worse it turns an issue about function into an issue about performance. I'm OK with this as a short fix.